### PR TITLE
added region folding

### PR DESCRIPTION
**Pull request by LucianoBestia**
Regions of code that you'd like to be folded can be wrapped with  `// #region` and `// #endregion` line comments.
This is called "Region Folding". It is originally available for many languages in VSCode. But Rust-analyzer has its own folding function and this is missing.
With this Pull Request I am suggesting a simple solution. 
The regions are a special kind of comments, so I added a bit of code in the comment folding function.
The regex to match are: `^\s*//\s*#?region\b` and `^\s*//\s*#?endregion\b`.
The number of space characters is not important. There is an optional # character. The line can end with a name of the region.
Example:
```rust
// 1. some normal comment
// region: test
// 2. some normal comment
calling_function(x,y);
// endregion: test
```
I added a test for this new functionality in `folding_ranges.rs`.
Please, take a look and comment. 
I found that these exact regexes are already present in the file `language-configuration.json`, but I don't find a way to read this configuration. So my regex is hardcoded in the code.

### CODE REVIEW

**Comment by matklad** [comment]
@dustypomerleau shouldn't this be solved by language configuration.json?

**Comment by LucianoBestia** [comment]
In the `language-configuration.json` there are these lines:
```
    "folding": {
        "markers": {
            "start": "^\\s*//\\s*#?region\\b",
            "end": "^\\s*//\\s*#?endregion\\b"
        }
    }
```
But the "region folding" does not work for rust files when using rust-analyzer.
I suppose that the original VSCode folding is not working, because rust-analyzer has its own folding functions.

**Comment by dustypomerleau** [comment]
`language-configuration.json` does normally fix this (and we are using the same folding regex as both TypeScript and Microsoft's [default Rust language config](https://github.com/microsoft/vscode-rust/blob/main/language-configuration.json).

However, after some tinkering it appears there is an RA-specific issue with folding regions that aren't present when the window loads. If you add `// #region` and `// #endregion` to a file with RA running, the folding arrow won't appear (and folding keybinding won't work) until you reload the window, presumably because this restarts the server. Needs a little investigating.

**Comment by matklad** [comment]
Yeah, re-reading the docs seems to hint that lanuguage-config.json does not work with server provided folds, so we need to do this manually. 

Implementation wise, I thing it's better to avoid regular expressions and to do all the required processing manually, using find and trim.

**Comment by LucianoBestia** [comment]
I used Regex to recognize regions because language-configuration.json contains these regexes.
We could instead just fix the format to simply: `// region:` and `// endregion`.

1. region is always declared inside a double slash comment. The code for folding starts inside the code for comments.
2. mandatory one space after double slash (no hash or multiple space or tabs allowed)
3. mandatory word `region` or `endregion`
4. optional colon to mark the beginning of the region name
5. optional what comes  after the colon is the region name

File: crates/ide/src/folding_ranges.rs [hunk 1]
```diff
@@ -48,10 +51,25 @@ pub(crate) fn folding_ranges(file: &SourceFile) -> Vec<Fold> {
...
+                        } else if comment.text().trim().starts_with("// endregion") {
+                            if !regions_starts.is_empty() {
+                                res.push(Fold {
+                                    range: TextRange::new(
+                                        regions_starts.pop().unwrap(),
+                                        comment.syntax().text_range().end(),
+                                    ),
+                                    kind: FoldKind::Region,
+                                })
+                            }
```
Comment (Veykril) [inline crates/ide/src/folding_ranges.rs:66]:
```suggestion
                            if let Some(region) = regions_starts.pop() {
                                res.push(Fold {
                                    range: TextRange::new(
                                        region,
                                        comment.syntax().text_range().end(),
                                    ),
                                    kind: FoldKind::Region,
                                })
                            }
```

File: crates/ide/src/folding_ranges.rs [hunk 2]
```diff
@@ -418,4 +444,24 @@ fn foo<fold arglist>(
 "#,
         )
     }
+
+    #[test]
+    fn fold_region() {
+        log_init_for_test_debug();
+        // only error level log is printed on the terminal
+        log::error!("test fold_region");
```
Comment (Veykril) [inline crates/ide/src/folding_ranges.rs:452]: I don't think the logging stuff is needed(including the `env_logger` dev-dependency)?

Comment (LucianoBestia) [reply crates/ide/src/folding_ranges.rs]: No, logging is not needed. I just wanted to have it temporarily.

**Comment by kjeremy** [comment]
If the language configuration only impacts vscode then shouldn't we just detect this on the client side? I find it a little odd that we're pushing a vscode specific convention down into the server.

**Comment by LucianoBestia** [comment]
`test check_code_formatting ... FAILED`
When I run it on my machine it passes. It is basically only `cargo fmt -- --check`.  
I use rustfmt 1.4.25-stable (0f29ff6 2020-11-11).
What version is used to check the pull request on GitHub?
Where can I find this info?

**Comment by lnicola** [comment]
I think we're using `cargo fmt` from the `stable` toolchain:

```
rustfmt 1.4.30-stable (8c6769d 2021-01-18)
```

There's a diff in the CI logs:

```diff
Diff in /home/runner/work/rust-analyzer/rust-analyzer/crates/parser/src/grammar/params.rs at line 169:
         let la1 = p.nth(1);
         let la2 = p.nth(2);
         let la3 = p.nth(3);
-        if !matches!((p.current(), la1, la2, la3),
-              (T![&], T![self], _, _)
-            | (T![&], T![mut], T![self], _)
-            | (T![&], LIFETIME_IDENT, T![self], _)
-            | (T![&], LIFETIME_IDENT, T![mut], T![self])
+        if !matches!(
+            (p.current(), la1, la2, la3),
+            (T![&], T![self], _, _)
+                | (T![&], T![mut], T![self], _)
+                | (T![&], LIFETIME_IDENT, T![self], _)
+                | (T![&], LIFETIME_IDENT, T![mut], T![self])
         ) {
             return m.abandon(p);
         }
Diff in /home/runner/work/rust-analyzer/rust-analyzer/crates/rust-analyzer/src/to_proto.rs at line 516:
     assert!(path.is_absolute());
     let url = lsp_types::Url::from_file_path(path).unwrap();
     match path.components().next() {
-        Some(path::Component::Prefix(prefix)) if matches!(prefix.kind(), path::Prefix::Disk(_) | path::Prefix::VerbatimDisk(_)) =>
+        Some(path::Component::Prefix(prefix))
+            if matches!(prefix.kind(), path::Prefix::Disk(_) | path::Prefix::VerbatimDisk(_)) =>
         {
             // Need to lowercase driver letter
         }
```

**Comment by LucianoBestia** [comment]
Thank you lnicola. 
I see the checks on Github are made with rustc 1.50.0, which comes with rustfmt 1.4.30-stable. 
I updated my machine to rustc 1.50.0 and run `cargo fmt`.
It changed some code, which is not my code, but the format must match.

**Comment by matklad** [comment]
bors r+

Thanks!

**Comment by lnicola** [comment]
regions2

**Comment by matklad** [comment]
Do we want to add regions to https://github.com/rust-analyzer/rust-analyzer/blob/13982e4ee42293862646d022273692d65a362885/crates/ide/src/file_structure.rs as well? Seems like we might want to .

**Pull request merged**